README.md
# motifcensus

> Count 3-nodes motifs and positions for unipartite networks.

[![R CMD Check](https://github.com/Ecosystem-Assessments/motifcensus/actions/workflows/main.yml/badge.svg)](https://github.com/Ecosystem-Assessments/motifcensus/actions/workflows/main.yml)
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)


## Description

So far, this package includes `motif_census()`, a function that counts the
different positions occupied by the different species in all the **3 species
motifs** of a given **unipartite network**. Also, `motif_census_triplet()` return
details for all triplets.

Note that `count_motif()` in [`igraph`](https://igraph.org/r/) counts motifs of
size 3 and 4 of different kind networks, but does not account for the different
positions occupied within a motif.


## Installation

```r
install.packages("remotes")
install_github("Ecosystem-Assessments/motifscensus")
```

## Example

```r
R> library(motifcensus)
R> # Network net: 4->3->1 & 4->3->2 & 1<-3->2 ; meaning 2 linear chains
R> # (motif 2 in Milo (2002)) and 1 exploitative exploitation (motif 1)
R> net <- rbind(c(0,0,0,0), c(0,0,0,0), c(1,1,0,0), c(0,0,1,0))
net
     [,1] [,2] [,3] [,4]
[1,]    0    0    0    0
[2,]    0    0    0    0
[3,]    1    1    0    0
[4,]    0    0    1    0

R> motif_census(net, unidirectional = TRUE)
$motifs
1 2 3 4 5
2 0 1 0 0

$motifs_node
     1 2 3 4 5
[1,] 1 0 1 0 0
[2,] 1 0 1 0 0
[3,] 2 0 1 0 0
[4,] 2 0 0 0 0

$positions
 [1] 2 2 2 0 0 1 2 0 0 0 0

$positions_node
     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11]
[1,]    0    0    1    0    0    0    1    0    0     0     0
[2,]    0    0    1    0    0    0    1    0    0     0     0
[3,]    0    2    0    0    0    1    0    0    0     0     0
[4,]    2    0    0    0    0    0    0    0    0     0     0


R> motif_census(net) # net is treated as bidirectional network
$motifs
 1  2  3  4  5  6  7  8  9 10 11 12 13
 1  2  0  0  0  0  0  0  0  0  0  0  0

$motifs_node
     1 2 3 4 5 6 7 8 9 10 11 12 13
[1,] 1 1 0 0 0 0 0 0 0  0  0  0  0
[2,] 1 1 0 0 0 0 0 0 0  0  0  0  0
[3,] 1 2 0 0 0 0 0 0 0  0  0  0  0
[4,] 0 2 0 0 0 0 0 0 0  0  0  0  0

$positions
 [1] 1 2 2 2 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0

$positions_node
     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11]
[1,]    0    1    0    0    1    0    0    0    0     0     0
[2,]    0    1    0    0    1    0    0    0    0     0     0
[3,]    1    0    0    2    0    0    0    0    0     0     0
[4,]    0    0    2    0    0    0    0    0    0     0     0
     [,12] [,13] [,14] [,15] [,16] [,17] [,18] [,19] [,20] [,21]
[1,]     0     0     0     0     0     0     0     0     0     0
[2,]     0     0     0     0     0     0     0     0     0     0
[3,]     0     0     0     0     0     0     0     0     0     0
[4,]     0     0     0     0     0     0     0     0     0     0
     [,22] [,23] [,24] [,25] [,26] [,27] [,28] [,29] [,30]
[1,]     0     0     0     0     0     0     0     0     0
[2,]     0     0     0     0     0     0     0     0     0
[3,]     0     0     0     0     0     0     0     0     0
[4,]     0     0     0     0     0     0     0     0     0

R> motif_census_triplet(net)
   n i j k i_to_j j_to_i i_to_k k_to_i j_to_k k_to_j pos_i pos_j pos_k sum_pos                 motif                 name_uni
1  0 0 1 3  FALSE  FALSE  FALSE  FALSE  FALSE  FALSE     0     0     0       0 i   j + i   k + j   k                         
2 34 0 2 3  FALSE   TRUE  FALSE  FALSE  FALSE   TRUE     5     4     3      12 i<--j + i   k + j<--k             linear chain
3 34 1 2 3  FALSE   TRUE  FALSE  FALSE  FALSE   TRUE     5     4     3      12 i<--j + i   k + j<--k             linear chain
4 40 0 1 2  FALSE  FALSE  FALSE   TRUE  FALSE   TRUE     2     2     1       5 i   j + i<--k + j<--k exploitative competition


R> head(bidirectional_motifs3)
  mid i_to_j j_to_i i_to_k k_to_i j_to_k k_to_j pid_i pid_j pid_k psum                 motif                 name_uni
1   0  FALSE  FALSE  FALSE  FALSE  FALSE  FALSE     0     0     0    0 i   j + i   k + j   k                         
2   1   TRUE  FALSE  FALSE  FALSE  FALSE  FALSE     0     0     0    0 i-->j + i   k + j   k                         
3   2  FALSE   TRUE  FALSE  FALSE  FALSE  FALSE     0     0     0    0 i<--j + i   k + j   k                         
4   3   TRUE   TRUE  FALSE  FALSE  FALSE  FALSE     0     0     0    0 i<->j + i   k + j   k                         
5   4  FALSE  FALSE   TRUE  FALSE  FALSE  FALSE     0     0     0    0 i   j + i-->k + j   k                         
6   5   TRUE  FALSE   TRUE  FALSE  FALSE  FALSE     1     2     2    5 i-->j + i-->k + j   k exploitative competition

```

NEWS.md
# devel 

* Data frame `table64` has been renamed `bidirectional_motifs3` and is now 
properly documented.
* `motif_census_all_triplets()` now only has one parameter, the number of rows is 
computed directly in the C++ function and unit tests have been added. 


# motifcensus 0.1.0

* Initial release, includes `motif_census()` and `motif_census_triplet()`.

tests/testthat.R
library(testthat)
library(motifcensus)

test_check("motifcensus")


tests/testthat/test_motif_census.R
# linear chain
mat_lc <- rbind(c(0, 1, 0), c(0, 0, 1), c(0, 0, 0))
mat_lc2 <- rbind(c(0, 1, 0), c(0, 0, 0), c(1, 0, 0))
# apparent competition
mat_ac <- rbind(c(0, 0, 0), c(1, 0, 0), c(1, 0, 0))
# exploitative competition
mat_ec <- rbind(c(0, 0, 0), c(0, 0, 0), c(1, 1, 0))
# omnivory
mat_om <- rbind(c(0, 0, 0), c(1, 0, 0), c(1, 1, 0))
mat_om2 <- rbind(c(0, 1, 1), c(0, 0, 1), c(0, 0, 0))
# circular
mat_ci <- rbind(c(0, 0, 1), c(1, 0, 0), c(0, 1, 0))
# 2 linear chain and one exploitative competition
mat_cb <- rbind(c(0, 0, 0, 0), c(0, 0, 0, 0), c(1, 1, 0, 0), c(0, 0, 1, 0))

test_that("expected error", {
  expect_error(motif_census(mat_lc[, -1], unidirectional = TRUE),
    "`mat` is not square",
    fixed = TRUE
  )
})


res_lc <- motif_census(t(mat_lc), unidirectional = TRUE)$positions_node
res_lc2 <- motif_census(mat_lc2, unidirectional = TRUE)$positions_node
res_ac <- motif_census(mat_ac, unidirectional = TRUE)$positions_node
res_ec <- motif_census(mat_ec, unidirectional = TRUE)$positions_node
res_om <- motif_census(mat_om, unidirectional = TRUE)$positions_node
res_om2 <- motif_census(mat_om2, unidirectional = TRUE)$positions_node
res_ci <- motif_census(mat_ci, unidirectional = TRUE)$positions_node
res_cb <- motif_census(mat_cb, unidirectional = TRUE)

test_that("position for unidirectional networks", {
  expect_equal(sum(res_lc), 3)
  expect_true(res_lc[1, 3] & res_lc[2, 2] & res_lc[3, 1])
  expect_equal(sum(res_lc2), 3)
  expect_true(res_lc2[3, 1] & res_lc2[1, 2] & res_lc2[2, 3])
  expect_equal(sum(res_ac), 3)
  expect_true(res_ac[1, 5] & res_ac[2, 4] & res_ac[3, 4])
  expect_equal(sum(res_ec), 3)
  expect_true(res_ec[1, 7] & res_ec[2, 7] & res_ec[3, 6])
  expect_equal(sum(res_om), 3)
  expect_true(res_om[1, 10] & res_om[2, 9] & res_om[3, 8])
  expect_equal(sum(res_om2), 3)
  expect_true(res_om2[1, 8] & res_om2[2, 9] & res_om2[3, 10])
  expect_equal(sum(res_ci), 3)
  expect_true(all(res_ci[, 11] == 1))

  cb_pos <- res_cb$positions_node
  expect_equal(sum(cb_pos), 9)
  expect_equal(sum(cb_pos[4, 1]), 2)
  expect_equal(sum(cb_pos[3, 2]), 2)
  expect_true(cb_pos[1, 3] & cb_pos[2, 3])
  expect_true(cb_pos[1, 7] & cb_pos[2, 7] & cb_pos[3, 6])

  expect_true(all(res_cb$positions == c(2, 2, 2, 0, 0, 1, 2, 0, 0, 0, 0)))
})

test_that("motifs for unidirectional networks", {
  expect_true(all(res_cb$species_motifs[, 1L] == c(1, 1, 2, 2)))
  expect_true(all(res_cb$motifs == c(2, 0, 1, 0, 0)))
})


test_that("positions for bidirectional networks", {
  expect_identical(motif_census(mat_ci), motif_census(t(mat_ci)))
})



test_that("triplet for 3 species networks", {
  expect_equal(motif_census_triplet(mat_lc)$mid, 17)
  expect_equal(motif_census_triplet(mat_lc2)$mid, 9)
  expect_equal(motif_census_triplet(mat_ac)$mid, 10)
  expect_equal(motif_census_triplet(mat_ec)$mid, 40)
  expect_equal(motif_census_triplet(mat_om)$mid, 42)
  expect_equal(motif_census_triplet(mat_om2)$mid, 21)
  expect_equal(motif_census_triplet(mat_ci)$mid, 38)
})

test_that("triplet for 4 species network", {
  expect_equal(motif_census_triplet(mat_cb)$mid, c(0, 34, 34, 40))
})

R/zzz.R
#' @importFrom utils data
#' @importFrom Rcpp evalCpp
#' @useDynLib motifcensus
#' @name motif_census
NULL

# Code used to generate bidirectional_motifs3

gen_bidirectional_motifs3 <- function() {
   out <- data.frame(
      mid = 0:63,
      i_to_j = FALSE, j_to_i = FALSE,
      i_to_k = FALSE, k_to_i = FALSE,
      j_to_k = FALSE, k_to_j = FALSE,
      motifcensus::positions_motif3,
      psum = NA_integer_,
      motif = NA_character_,
      name_uni = "",
      stringsAsFactors = FALSE
   )
   names(out)[8:10] <- paste0("pid_", names(out)[8:10])
   # the sum below identify motif
   out$psum <- apply(out[8:10], 1, sum)

   # detail what interaction are on
   for (i in 0:63) {
      for (j in 0:5) {
         if (bitwAnd(bitwShiftR(i, j), 1)) out[i + 1, j + 2] <- TRUE
      }
      out$motif[i + 1] <- paste(
         write_ntw(out[i + 1, 2] + 2 * out[i + 1, 3] + 1),
         write_ntw(out[i + 1, 4] + 2 * out[i + 1, 5] + 1, "i", "k"),
         write_ntw(out[i + 1, 6] + 2 * out[i + 1, 7] + 1, "j", "k"),
         sep = " + "
      )
   }

   # motifs name (unidirectional version)
   out$name_uni[out$psum == 5] <- "exploitative competition"
   out$name_uni[out$psum == 12] <- "linear chain"
   out$name_uni[out$psum == 28] <- "apparent competition"
   out$name_uni[out$psum == 36] <- "omnivory"

   out
}
# bidirectional_motifs3 <- gen_bidirectional_motifs3()
# save(bidirectional_motifs3, file = "data/bidirectional_motifs3.rda")

write_ntw <- function(val, x = "i", y = "j") {
   arr <- switch(val,
      "   ",
      "-->",
      "<--",
      "<->"
   )
   paste0(x, arr, y)
}


R/data.R
#' potisions_motifs
#'
#' Table giving the motifs positions associated with all the 64 combinations
#' of interactions for 3 species motifs.
#'
#' @docType data
#' @keywords datasets
#' @name positions_motif3
#' @usage positions_motif3
#' @format a matrix of 64 rows and 3 columns
"positions_motif3"


#' Table of motifs
#'
#' Display all the 64 bidirectional 3-species motifs, for every motif,  details interactions and name notable motifs.
#'
#' @docType data
#' @keywords datasets
#' @name bidirectional_motifs3
#' @usage bidirectional_motifs3
#' @references
#' * Milo, R. (2002). Network Motifs: Simple Building Blocks of Complex Networks. Science, 298(5594), 824–827. https://doi.org/10.1126/science.298.5594.824
#' @format a matrix of 64 rows (1 row per motif) and 13 columns: 
#' * `mid`: motif identifier
#' * `i_to_j`: a logical, is there an interaction from `i` to ``j``;
#' * `j_to_i`: a logical, is there an interaction from `j` to ``i``;
#' * `i_to_k`: a logical, is there an interaction from `i` to ``k``;
#' * `k_to_i`: a logical, is there an interaction from `k` to ``i``;
#' * `j_to_k`: a logical, is there an interaction from `j` to ``k``;
#' * `k_to_j`: a logical, is there an interaction from `k` to ``j``;
#' * `pid_i`: position identifier of `i`;
#' * `pid_j`: position identifier of `j`;
#' * `pid_k`: position identifier of `k`;
#' * `psum`: sum of positions (could be used to identify motifs);
#' * `motif`: motifs description using ascii characters;
#' * `name_uni`: motif name (in unidirectionnal networks).
"bidirectional_motifs3"

R/motif_census.R
#' Compute 3-nodes motif census
#'
#' Compute 3-nodes motif census for unidirection and bidirectional
#' unipartite networks.
#'
#' @param mat a square matrix of logical describing a network.
#' @param unidirectional a logical. Is the network considered unidirectional?
#' Default is set to `FALSE` meaning that the network is bidirectional.
#'
#' @references
#' * Milo, R. (2002). Network Motifs: Simple Building Blocks of Complex
#' Networks. Science, 298(5594), 824–827. https://doi.org/10.1126/science.298.
#' 5594.824
#'
#' @export
#'
#' @return
#' A list of two elements:
#' * `motifs`: total motif count.
#' * `motifs_node`: motif count per node.
#' * `positions`: total position count.
#' * `positions_node`: position count per nodes
#'
#' If `unidirectional` is `TRUE`, then motifs ordered as follows:
#'  * 1 linear chains bottom;
#' 	* 2 apparent competition;
#' 	* 3 exploitative competition;
#' 	* 4 omnivory;
#' 	* 5 circular;
#' and positions are as follows:
#' 	* 1 linear chains bottom;
#' 	* 2 linear chains middle;
#' 	* 3 linear chains top;
#' 	* 4 apparent competition botton;
#' 	* 5 apparent competition top;
#' 	* 6 exploitative competition bottom;
#' 	* 7 exploitative competition top;
#' 	* 8 omnivory bottom;
#' 	* 9 omnivory middle;
#' 	* 10 omnivory top;
#' 	* 11 circular (e.g. 1->2->3->1).
#'
#' If `mat` corresponds to a bidirectional network, then the 13 motifs are
#' ordered following Milo (2002), see the documentation of
#' [motifcensus::bidirectional_motifs3] for more details.
#' @examples
#' net <- rbind(c(0, 0, 0, 0), c(0, 0, 0, 0), c(1, 1, 0, 0), c(0, 0, 1, 0))
#' motif_census(net, unidirectional = TRUE)
#' motif_census_triplet(net)
motif_census <- function(mat, unidirectional = FALSE) {
    if (unidirectional) {
        pos <- motif_census_uni(mat)
        pos2mot <- rep(seq_len(5), c(3, 2, 2, 3, 1))
    } else {
        pos <- motif_census_bi(mat, motifcensus::positions_motif3)
        pos2mot <- rep(seq_len(13), c(2, 3, 3, 2, 3, 2, 3, 2, 1, 3, 2, 3, 1))
    }
    tmp_mot <- split(t(pos), pos2mot)
    mot <- unlist(lapply(tmp_mot, function(x) sum(x) / 3))
    motsp <- motifs_node(tmp_mot, nrow(mat))
    list(
        motifs = mot,
        motifs_node = motsp,
        positions = apply(pos, 2, sum),
        positions_node = pos
    )
}

#' @export
#' @describeIn motif_census return details for all motifs
motif_census_triplet <- function(mat) {
    out <- as.data.frame(motif_census_all_triplets(mat))
    names(out) <- c("i", "j", "k", "mid")
    merge(out, motifcensus::bidirectional_motifs3)
}


motifs_node <- function(x, nsp) {
    t(do.call(
        rbind,
        lapply(x, function(y) apply(matrix(y, ncol = nsp), 2, sum))
    ))
}

src/RcppExports.cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include <Rcpp.h>

using namespace Rcpp;

#ifdef RCPP_USE_GLOBAL_ROSTREAM
Rcpp::Rostream<true>&  Rcpp::Rcout = Rcpp::Rcpp_cout_get();
Rcpp::Rostream<false>& Rcpp::Rcerr = Rcpp::Rcpp_cerr_get();
#endif

// motif_census_uni
NumericMatrix motif_census_uni(LogicalMatrix mat);
RcppExport SEXP _motifcensus_motif_census_uni(SEXP matSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< LogicalMatrix >::type mat(matSEXP);
    rcpp_result_gen = Rcpp::wrap(motif_census_uni(mat));
    return rcpp_result_gen;
END_RCPP
}
// motif_census_bi
NumericMatrix motif_census_bi(LogicalMatrix mat, IntegerMatrix ref);
RcppExport SEXP _motifcensus_motif_census_bi(SEXP matSEXP, SEXP refSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< LogicalMatrix >::type mat(matSEXP);
    Rcpp::traits::input_parameter< IntegerMatrix >::type ref(refSEXP);
    rcpp_result_gen = Rcpp::wrap(motif_census_bi(mat, ref));
    return rcpp_result_gen;
END_RCPP
}
// motif_census_all_triplets
NumericMatrix motif_census_all_triplets(LogicalMatrix mat);
RcppExport SEXP _motifcensus_motif_census_all_triplets(SEXP matSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< LogicalMatrix >::type mat(matSEXP);
    rcpp_result_gen = Rcpp::wrap(motif_census_all_triplets(mat));
    return rcpp_result_gen;
END_RCPP
}

static const R_CallMethodDef CallEntries[] = {
    {"_motifcensus_motif_census_uni", (DL_FUNC) &_motifcensus_motif_census_uni, 1},
    {"_motifcensus_motif_census_bi", (DL_FUNC) &_motifcensus_motif_census_bi, 2},
    {"_motifcensus_motif_census_all_triplets", (DL_FUNC) &_motifcensus_motif_census_all_triplets, 1},
    {NULL, NULL, 0}
};

RcppExport void R_init_motifcensus(DllInfo *dll) {
    R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
    R_useDynamicSymbols(dll, FALSE);
}


src/motif_census_core.cpp

#include <Rcpp.h>
using namespace Rcpp;

// [[Rcpp::export]]
NumericMatrix motif_census_uni(LogicalMatrix mat)
{

	if (mat.nrow() != mat.ncol())
	{ // log() not defined here
		stop("`mat` is not square");
	}

	NumericMatrix out(mat.nrow(), 11);
	// 11 unique positions
	// 0-2 => linear chains => 0- bottom 1- medium 2- top
	// 3-4 => apparent competition => 3- bottom 4- top
	// 5-6 => exploitative competition => 5- bottom 6- top
	// 7-9 => omnivory => 7- bottom 8- medium -9 top
	// 10 => circular
	int j, k, i;
	int nsp = mat.nrow();

	// REMOVE bidirectional interactions and cannibalism
	for (i = 0; i < mat.nrow(); i++)
	{
		mat(i, i) = false;
		for (j = i + 1; j < mat.ncol(); j++)
		{
			if (mat(i, j) && mat(j, i))
			{
				mat(i, j) = false;
				mat(j, i) = false;
			}
		}
	}

	// Here starts the part where interactions are checked
	for (i = 0; i < nsp - 2; i++)
	{
		for (j = i + 1; j < nsp - 1; j++)
		{
			for (k = j + 1; k < nsp; k++)
			{
				// there are 27 possibilities
				if (mat(j, i))
				{ // j -> i
					//
					if (mat(k, j))
					{ // k -> j
						if (mat(k, i))
						{
							out(i, 9)++;
							out(j, 8)++;
							out(k, 7)++;
						}
						else if (mat(i, k))
						{
							out(i, 10)++;
							out(j, 10)++;
							out(k, 10)++;
						}
						else
						{
							out(i, 2)++;
							out(j, 1)++;
							out(k, 0)++;
						}
					}
					else if (mat(j, k))
					{ // j -> k
						if (mat(k, i))
						{
							out(i, 9)++;
							out(j, 7)++;
							out(k, 8)++;
						}
						else if (mat(i, k))
						{
							out(i, 8)++;
							out(j, 7)++;
							out(k, 9)++;
						}
						else
						{
							out(i, 6)++;
							out(j, 5)++;
							out(k, 6)++;
						}
					}
					else
					{
						if (mat(k, i))
						{
							out(i, 4)++;
							out(j, 3)++;
							out(k, 3)++;
						}
						else if (mat(i, k))
						{
							out(i, 1)++;
							out(j, 0)++;
							out(k, 2)++;
						}
						else
						{
							// no motif
						}
					}
				}
				else if (mat(i, j))
				{ // i -> j
					//
					if (mat(k, j))
					{
						if (mat(k, i))
						{
							out(i, 8)++;
							out(j, 9)++;
							out(k, 7)++;
						}
						else if (mat(i, k))
						{
							out(i, 7)++;
							out(j, 9)++;
							out(k, 8)++;
						}
						else
						{
							out(i, 3)++;
							out(j, 4)++;
							out(k, 3)++;
						}
					}
					else if (mat(j, k))
					{
						if (mat(k, i))
						{
							out(i, 10)++;
							out(j, 10)++;
							out(k, 10)++;
						}
						else if (mat(i, k))
						{
							out(i, 7)++;
							out(j, 8)++;
							out(k, 9)++;
						}
						else
						{
							out(i, 0)++;
							out(j, 1)++;
							out(k, 2)++;
						}
					}
					else
					{
						if (mat(k, i))
						{
							out(i, 1)++;
							out(j, 2)++;
							out(k, 0)++;
						}
						else if (mat(i, k))
						{
							out(i, 5)++;
							out(j, 6)++;
							out(k, 6)++;
						}
						else
						{
							// no motif
						}
					}
				}
				else
				{
					if (mat(k, j))
					{
						if (mat(k, i))
						{
							out(i, 6)++;
							out(j, 6)++;
							out(k, 5)++;
						}
						else if (mat(i, k))
						{
							out(i, 0)++;
							out(j, 2)++;
							out(k, 1)++;
						}
						else
						{
							// no motif
						}
					}
					else if (mat(j, k))
					{
						if (mat(k, i))
						{
							out(i, 2)++;
							out(j, 0)++;
							out(k, 1)++;
						}
						else if (mat(i, k))
						{
							out(i, 3)++;
							out(j, 3)++;
							out(k, 4)++;
						}
						// no motif otherwise (so else is droped
					}
					// no motif otherwise (so else is droped)
				}
			}
		}
	}
	return out;
}



// [[Rcpp::export]]
NumericMatrix motif_census_bi(LogicalMatrix mat, IntegerMatrix ref)
{

	if (mat.nrow() != mat.ncol())
	{ // log() not defined here
		stop("`mat` is not square");
	}

	// 30 unique positions
	NumericMatrix out(mat.nrow(), 30);
	// NB: ordered following the 13 motifs described in Milo (2002)
	int j, k, i, s, ss;
	int nsp = mat.nrow();

	// REMOVE cannibalism
	for (i = 0; i < mat.nrow(); i++)
		mat(i, i) = false;

	// count motifs
	for (i = 0; i < nsp - 2; i++)
	{
		for (j = i + 1; j < nsp - 1; j++)
		{
			s = mat(i, j) ? 1 : 0;  // i->j
			s += mat(j, i) ? 2 : 0; // j->i
			for (k = j + 1; k < nsp; k++)
			{
				ss = s;
				ss += mat(i, k) ? 4 : 0;  // i->k
				ss += mat(k, i) ? 8 : 0;  // k->i
				ss += mat(j, k) ? 16 : 0; // j->k
				ss += mat(k, j) ? 32 : 0; // k->j
				// if first is different of 0
				if (ref(ss, 0))
				{
					out(i, ref(ss, 0)-1)++;
					out(j, ref(ss, 1)-1)++;
					out(k, ref(ss, 2)-1)++;
				}
			}
		}
	}
	return out;
}



// [[Rcpp::export]]
NumericMatrix motif_census_all_triplets(LogicalMatrix mat)
{

		if (mat.nrow() != mat.ncol())
		{ // log() not defined here
				stop("`mat` is not square");
		}

		NumericMatrix out(Rf_choose(mat.nrow(), 3), 4);
		// NB: ordered following the 13 motifs described in Milo (2002)
		int j, k, i, r, s, ss;
		int nsp = mat.nrow();

		// REMOVE cannibalism
		for (i = 0; i < mat.nrow(); i++)
				mat(i, i) = false;

		// count motifs
		r = 0;
		for (i = 0; i < nsp - 2; i++)
		{
				for (j = i + 1; j < nsp - 1; j++)
				{
						s = mat(i, j) ? 1 : 0; // i->j
						s += mat(j, i) ? 2 : 0; // j->i
						for (k = j + 1; k < nsp; k++)
						{
								ss = s;
								ss += mat(i, k) ? 4 : 0; // i->k
								ss += mat(k, i) ? 8 : 0; // k->i
								ss += mat(j, k) ? 16 : 0; // j->k
								ss += mat(k, j) ? 32 : 0; // k->j
								// edit out
								out(r, 0) = i;
								out(r, 1) = j;
								out(r, 2) = k;
								out(r, 3) = ss;
								r++;
						}
				}
		}
		return out;
}

inst/data/create_table.R
positions_motif3 <- matrix(0, 64, 3)

# Below 1 + xx is used to bear in mind the converstion between c++/R

# **motif 1** expoitative competition, 3 cases
# 2 positions -- bottom (1): 1 and top (2): 2
positions_motif3[1 + 5,] <- c(1, 2, 2)
positions_motif3[1 + 18,] <- c(2, 1, 2)
positions_motif3[1 + 40,] <- c(2, 2, 1)

# **motif 2** linear chain, 6 cases
# 3 positions -- bottom: 3, middle: 4 and top: 5
positions_motif3[1 + 6,] <- c(4, 3, 5) # j->i->k
positions_motif3[1 + 9,] <- c(4, 5, 3) # k->i->j
positions_motif3[1 + 17,] <- 3:5 # i->j->k
positions_motif3[1 + 24,] <- c(5, 3, 4) # j->k->i
positions_motif3[1 + 34,] <- c(5, 4, 3) # k->j->i
positions_motif3[1 + 36,] <- c(3, 5, 4) # i->k->j 

# **motif 3** A->B->C + B->A, 6 cases
# 3 positions -- bottom: 6, middle: 7 and top: 8
positions_motif3[1 + 7 ,] <- c(7, 6, 8) # j<=>i->k
positions_motif3[1 + 13,] <- c(7, 8, 6) # k<=>i->j 
positions_motif3[1 + 19,] <- c(6, 7, 8) # i<=>j->k 
positions_motif3[1 + 56,] <- c(8, 6, 7) # j<=>k->i 
positions_motif3[1 + 50,] <- c(8, 7, 6) # k<=>j->i 
positions_motif3[1 + 44,] <- c(6, 8, 7) # i<=>k->j

# **motif 4** apparent competition, 3 cases 
# 2 positions -- bottom (2): 9 and top (1): 10
positions_motif3[1 + 10,] <- c(10, 9, 9) # k->i<-j
positions_motif3[1 + 20,] <- c(9, 9, 10) # i->k<-j
positions_motif3[1 + 33,] <- c(9, 10, 9) # k->j<-i

# **motif 5** omnivory, 6 cases 
# 3 positions -- bottom: 11 middle: 12 top: 13
positions_motif3[1 + 21,] <- c(11, 12, 13) # i->k<-j + j->i
positions_motif3[1 + 22,] <- c(12, 11, 13) # i->k<-j + i->j 
positions_motif3[1 + 26,] <- c(13, 11, 12) # j->i<-k + j->k
positions_motif3[1 + 37,] <- c(11, 13, 12) # k->j<-j + i->k
positions_motif3[1 + 41,] <- c(12, 13, 11) # k->j<-i + k->i
positions_motif3[1 + 42,] <- c(13, 12, 11) # k->i<-j + k->j

# **motif 6** omnivory, 3 cases 
# 2 positions -- bottom (2): 14 top (2): 15
positions_motif3[1 + 23,] <- c(14, 14, 15) # i->k<-j + j<=>i
positions_motif3[1 + 45,] <- c(14, 15, 14) # i->j<-k + i<=>k 
positions_motif3[1 + 58,] <- c(15, 14, 14) # j->i<-k + j<=>k


# **motif 7** A <=> B => C, 3 cases 
# 3 positions -- 3 links: 16,  2 links: 17, 1 link:: 18
positions_motif3[1 + 11,] <- c(16, 17, 18) # j<=>i<- k
positions_motif3[1 + 14,] <- c(16, 18, 17) # k<=>i<- j 
positions_motif3[1 + 28,] <- c(17, 18, 16) # i<=>k<- j
positions_motif3[1 + 35,] <- c(17, 16, 18) # i<=>j<- k
positions_motif3[1 + 49,] <- c(18, 16, 17) # k<=>j<- i
positions_motif3[1 + 52,] <- c(18, 17, 16) # j<=>k<- i


# **motif 8** A <=> B <=> C, 3 cases 
# 2 positions -- 4 links: 19,  2 links: 20
positions_motif3[1 + 15,] <- c(19, 20, 20) # j<=>i<=>k
positions_motif3[1 + 51,] <- c(20, 19, 20) # i<=>j<=>k 
positions_motif3[1 + 60,] <- c(20, 20, 19) # j<=>k<=>i


# **motif 9** circluar, 2 cases
# 1 unique position -- 21
positions_motif3[1 + c(25, 38),] <- 21


# **motif 10** A -> B <=> C + C -> A
# 3 positions -- A: 22 B: 23 C: 24 (did not find a good way to name those)
positions_motif3[1 + 27,] <- c(23, 24, 22) # k->i<=>j + j->k
positions_motif3[1 + 29,] <- c(24, 22, 23) # j->k<=>i + i->j 
positions_motif3[1 + 39,] <- c(24, 23, 22) # k->j<=>i + i->k
positions_motif3[1 + 46,] <- c(23, 22, 24) # j->i<=>k + k->j
positions_motif3[1 + 54,] <- c(22, 24, 23) # i->k<=>j + j->i
positions_motif3[1 + 57,] <- c(22, 23, 24) # i->j<=>k + k->i


# **motif 11**  C <- A -> B + B <=> C, 3 cases
# 2 positions -- bottom (1): 25  top (2): 26 
positions_motif3[1 + 30,] <- c(26, 25, 26) # i<-j->k + i<=>k
positions_motif3[1 + 43,] <- c(26, 26, 25) # i<-k->j + i<=>j
positions_motif3[1 + 53,] <- c(25, 26, 26) # j<-i->k + j<=>k


# **motif 12**  A <=> B <=> C + A -> C
# 3 positions -- B: 27 A: 28 A: 29 (did not find a good way to name those)
positions_motif3[1 + 31,] <- c(27, 29, 28) # k<=>i<=>j + j->k 
positions_motif3[1 + 47,] <- c(27, 28, 29) # j<=>i<=>k + k->j  
positions_motif3[1 + 55,] <- c(29, 27, 28) # k<=>j<=>i + i->j 
positions_motif3[1 + 59,] <- c(28, 27, 29) # i<=>j<=>k + k->i 
positions_motif3[1 + 61,] <- c(29, 28, 27) # j<=>k<=>i + i->j 
positions_motif3[1 + 62,] <- c(28, 29, 27) # i<=>k<=>j + j->i 


## **motif 13** fully connected, 1 case
## position: 30
positions_motif3[1 + 63,] <- 30
colnames(positions_motif3) <- c("i", "j", "k")


# write.csv(positions_motif3, file = "data/positions_motif3.csv", 
#     row.names = FALSE)
save(positions_motif3, file="data/positions_motif3.rdata", 
    compress = "xz", version = 2)
